File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/game/camera/FieldJournalViewModel.kt
package com.jworks.kanjiquest.android.ui.game.camera

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.core.domain.model.FieldJournalEntry
import com.jworks.kanjiquest.core.domain.repository.FieldJournalRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class FieldJournalUiState(
    val entries: List<FieldJournalEntry> = emptyList(),
    val totalPhotos: Long = 0,
    val totalKanjiCaught: Long = 0,
    val isLoading: Boolean = true,
    val selectedEntry: FieldJournalEntry? = null
)

@HiltViewModel
class FieldJournalViewModel @Inject constructor(
    private val repository: FieldJournalRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(FieldJournalUiState())
    val uiState: StateFlow<FieldJournalUiState> = _uiState.asStateFlow()

    init {
        loadEntries()
    }

    private fun loadEntries() {
        viewModelScope.launch {
            val entries = repository.getAll()
            val total = repository.countAll()
            val kanjiCaught = repository.totalKanjiCaught()
            _uiState.value = FieldJournalUiState(
                entries = entries,
                totalPhotos = total,
                totalKanjiCaught = kanjiCaught,
                isLoading = false
            )
        }
    }

    fun selectEntry(entry: FieldJournalEntry) {
        _uiState.value = _uiState.value.copy(selectedEntry = entry)
    }

    fun clearSelection() {
        _uiState.value = _uiState.value.copy(selectedEntry = null)
    }

    fun deleteEntry(id: Long) {
        viewModelScope.launch {
            repository.delete(id)
            loadEntries()
            _uiState.value = _uiState.value.copy(selectedEntry = null)
        }
    }
}

File: ios-app/KanjiQuest/Screens/Game/Camera/FieldJournalViewModel.swift
import Foundation
import SharedCore

@MainActor
final class FieldJournalViewModel: ObservableObject {
    @Published var entries: [FieldJournalEntry] = []
    @Published var totalPhotos: Int64 = 0
    @Published var totalKanjiCaught: Int64 = 0
    @Published var isLoading = true
    @Published var selectedEntry: FieldJournalEntry? = nil

    private var repository: FieldJournalRepository?

    func load(container: AppContainer) {
        repository = container.fieldJournalRepository
        loadEntries()
    }

    private func loadEntries() {
        Task {
            let all = (try? await repository?.getAll()) ?? []
            let totalRaw = try? await repository?.countAll()
            let caughtRaw = try? await repository?.totalKanjiCaught()

            entries = all
            totalPhotos = totalRaw?.int64Value ?? 0
            totalKanjiCaught = caughtRaw?.int64Value ?? 0
            isLoading = false
        }
    }

    func selectEntry(_ entry: FieldJournalEntry) {
        selectedEntry = entry
    }

    func clearSelection() {
        selectedEntry = nil
    }

    func deleteEntry(_ id: Int64) {
        Task {
            try? await repository?.delete(id: id)
            loadEntries()
            selectedEntry = nil
        }
    }
}
